README.md
# GuardUniform
GuardUniform Bukkit Plugin for MC-Convicted.org

## convictedisdead
- [The Convicted Minecraft Server officially died on November 24, 2015.](http://www.minecraftforum.net/forums/servers/pc-servers/765507-convicted-classic-prison-server-non-op?comment=65094)
- Developed and tested using the Bukkit API, CraftBukkit Server, and Minecraft Client versions within June 2012 - July 2014
- This was my second and last plugin for the server

## Dedications
I dedicate this to ViperKnows, who offered me an amazing opportunity to be a part of his server's dev team.


changelog.txt
GuardUniform 1.73 Documentation

1. Guards will always be wearing chain armor with a certain enchantment
depending on their rank. This chain armor cannot be taken off, and will not drop on death.

The enchantments of the armor are configurable in config.yml.
The armor permission nodes are as follows:

uniform.alphaguard
uniform.guard
uniform.trainee

2. When a player with any of the above permission nodes joins the game, the plugin will check
if the player has the armor specified in config.yml.

If the player does not have the armor, then the plugin will try to auto-equip the new armor
and put the player's currently worn armor into his/her inventory.

If the player's inventory is full, the player will be sent a message telling them to free up
inventory space and relog to receive their uniform.

3. Players with the permission nodes specified above will also drop a full set of iron armor.
The durability of each of the armor pieces are calculated like so:

50% chance of the armor being 50% - 100% damaged
50% chance of the armor being 0 - 50% damaged

4. Players with the permission nodes specified above cannot drop health potions from their inventory
or drop them on death.

- This applies to both regular instant health (drinkable type) and splash instant health.

They will not be able to drop the potion whether it be by:
- Using the drop button key
- Removing it through the use of the inventory interface

5. If ever non-guards receive chain armor, they will not be able to equip it. 

me/menexia/guarduniform/GuardUniform.java
package me.menexia.guarduniform;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.logging.Level;
import java.util.logging.Logger;

import org.bukkit.Bukkit;
import org.bukkit.enchantments.Enchantment;
import org.bukkit.entity.Player;
import org.bukkit.plugin.java.JavaPlugin;

/*
[1:47:17 PM] Will: Guards will always be wearing chain armor with a certain enchantment
[1:47:26 PM] Will: depending on their rank
[1:47:29 PM] Will: trainee-guard-alphaguard
[1:47:35 PM] Will: this chain armor cannot be taken off
[1:47:50 PM] Will: and will not drop on death
[1:47:54 PM] Will: or be able to be stored in chests, etc.
DONE

[1:48:39 PM] Will: this armor should override any worn armor
[1:48:59 PM] Will: but would it be possible to take the armor theyre wearing and put it in their inventory?
[1:49:03 PM] Will: when they first become guard
[1:49:08 PM] Will: so they dont lose it
[1:50:04 PM] xavi ablaza: sure, what plugin does the server use for managing ranks?
[1:50:27 PM] Will: PEX
[1:50:36 PM] Will: a simple permission node would suffice
[1:50:41 PM] Will: say, convicted.guard
[1:50:43 PM] Will: or something
DONE

[1:52:08 PM] Will: yea
[1:52:20 PM] Will: so, even though guards will be forced to wear chain armor
[1:52:21 PM] Will: always
[1:52:30 PM] Will: I don't want other players to have it
[1:52:47 PM] Will: so anyone that does not have the guard node
[1:52:51 PM] Will: or is not a guard
[1:52:58 PM] Will: cannot equip chain armor
[1:53:07 PM] Will: or, if its easier, just delete it from the inventories of non-guards
[1:53:16 PM] Will: if they ever get ahold of it somehow
[1:53:32 PM] Will: main goal is to easily distinguish guards from the crowd
the armor can't be taken out in the first place, so non-guards can't have it all together.
DONE, POSSIBLY...

[1:55:35 PM] Will: so, the following should only happen to players with guard permission node
[1:55:49 PM] Will: when they die, they drop a full set of iron armor, as if they were wearing it
[1:55:57 PM] Will: (so there's still an incentive to kill guards)
[1:56:04 PM] Will: however, this armor has random damage to it
percentage like dynafish, 
50/50 chance for bottom half and top half
do a random chancing of the bottom half and the top half.
you can combine helmets together, also.

[1:57:04 PM] Will: ppl with guard permission node
[1:57:17 PM] Will: cannot remove health potions from their inventory
[1:57:21 PM] Will: or drop them on death
[1:57:30 PM] Will: they will get them via /kit, which you dont need to worry about
[1:57:38 PM] Will: but once they get them, they can't get rid of them
[1:57:40 PM] Will: unless they drink them
[1:57:49 PM] Will: or they die (in which case it just disappears)
[1:58:09 PM] Will: sorry if this is complicated. but thats about it :P
[1:58:15 PM] xavi ablaza: also for health splash?
[1:58:31 PM] Will: hmm
[1:58:37 PM] Will: yes
[1:58:42 PM] Will: health splash and drink
[1:58:51 PM] xavi ablaza: alright.

basically, health and splash potions also cannot be removed from a guard's inventory
or can be dropped on death

these pots can be arrangeable though, so probably check if the click range is null, or something

ADDITIONAL SHIZNITS
Changing enchant system, check if the armor was chain perhaps
Chain armor not to drop on death in general
When upgraded, change their armor
Fix chest stuff
 */
/**
 * 
 * @author Xavier Luis Ablaza - MeneXia
 *
 */
public class GuardUniform extends JavaPlugin {
    // Logger for console
    public Logger logger;
    
    // ArrayList of Maps containing Armor Enchantments
    // The Map esentially is a list of Enchantments for a particular armor piece
    public List<Map<Enchantment, Integer>> armorEnchants = new ArrayList<Map<Enchantment, Integer>>();
    
    // List of Healing Potions
    public final static Short[] healPots = { (short) 16341, (short) 16373, (short) 32725,
            (short) 32757 };

    @Override
    public void onDisable() {
        this.logger.info("Version " + this.getDescription().getVersion() + " disabled!");
    }

    @Override
    public void onEnable() {
        // Get logger
        logger = this.getLogger();
        
        // Generate config file
        checkConfig();
        
        // Loads non-volatile enchantments in config.yml into volatile array (optimization)
        loadEnchants();
        
        // Create new main listener and register its events
        new GUMainListener(this);

        // Check if config.yml states that it doesn't want chain armor to break. Ever.
        // Create new damage listener and register its events if yes.
        if (this.getConfig().getBoolean("noBreak", true)) {
            new GUDamageListener(this);
        }

        this.logger.info("Version " + this.getDescription().getVersion() + " enabled!");
    }

    /**
     * Generates config.yml file
     */
    private void checkConfig() {
        String name = "config.yml";
        File actual = new File(getDataFolder(), name);
        if (!actual.exists()) {
            getDataFolder().mkdir();
            InputStream input = getClass().getResourceAsStream("/defaults/config.yml");
            if (input != null) {
                FileOutputStream output = null;

                try {
                    output = new FileOutputStream(actual);
                    byte[] buf = new byte[4096]; // [8192]?
                    int length = 0;
                    while ((length = input.read(buf)) > 0) {
                        output.write(buf, 0, length);
                    }
                    this.logger.info("Default configuration file written: " + name);
                } catch (IOException e) {
                    e.printStackTrace();
                } finally {
                    try {
                        if (input != null)
                            input.close();
                    } catch (IOException e) {
                    }

                    try {
                        if (output != null)
                            output.close();
                    } catch (IOException e) {
                    }
                }
            }
        }
    }

    /**
     * Loads enchants of AlphaGuard, Guard, and Trainee from config.yml
     * into the Armor Enchantments ArrayList (volatile)
     * 
     * INDEX IS IMPORTANT!
     * 0 - AlphaGuard Helmet
     * 1 - AlphaGuard Chestplate
     * 2 - AlphaGuard Leggings 
     * 3 - AlphaGuard Boots 
     * 4 - Guard Helmet
     * 5 - Guard Chestplate 
     * 6 - Guard Leggings 
     * 7 - Guard Boots
     * 8 - Trainee Helmet
     * 9 - Trainee Chestplate 
     * 10 - Trainee Leggings 
     * 11 - Trainee Boots
     */
    public void loadEnchants() {
        final String base = "ArmorEnchants.";
        final String[] paths = {
                base + "AlphaGuard.HELM",
                base + "AlphaGuard.LEGGINGS",
                base + "AlphaGuard.CHESTPLATE",
                base + "AlphaGuard.BOOTS",
                base + "Guard.HELM",
                base + "Guard.CHESTPLATE",
                base + "Guard.LEGGINGS",
                base + "Guard.BOOTS",
                base + "Trainee.HELM",
                base + "Trainee.CHESTPLATE",
                base + "Trainee.LEGGINGS",
                base + "Trainee.BOOTS"
        };
        boolean hasErrors = false;
        
        // For every armor piece, get its armor enchants from the config.yml
        for (int i=0; i<12; i++) {
            armorEnchants.add(getArmorEnchantments(paths[i], hasErrors));
        }
        
        // If there is an error in the config.yml, disable the plugin
        if (hasErrors) {
            logger.log(Level.SEVERE, "Version " + this.getDescription().getVersion() + 
                    " disabled due to errors in config.yml (check above)!");
            Bukkit.getPluginManager().disablePlugin(this);
        }
    }

    /**
     * Get the specific enchantments for a specific armor piece
     * @param path to armor piece in config.yml
     * @param hasErrors flag variable
     * @return Map<Enchantment, Integer> of enchants in an armor piece
     */
    private Map<Enchantment, Integer> getArmorEnchantments(String path, boolean hasErrors) {
        Map<Enchantment, Integer> armorEnchants = new HashMap<Enchantment, Integer>();
        for (String enchant : this.getConfig().getStringList(path)) {
            String[] parts = enchant.trim().split(" ");
            String enchantName = parts[0]
            /*
             * .replaceAll("PROJECTILEPROTECTION", "PROTECTION_PROJECTILE")
             * .replaceAll("FIREPROTECTION", "PROTECTION_FIRE")
             * .replaceAll("BLASTPROTECTION", "PROTECTION_EXPLOSIONS")
             * .replaceAll("PROTECTION", "PROTECTION_ENVIRONMENTAL")
             * .replaceAll("FEATHERFALLING", "PROTECTION_FALL")
             * .replaceAll("RESPIRATION", "OXYGEN") .replaceAll("AQUAAFFINITY",
             * "WATER_WORKER")
             */;
            if (enchantName.equalsIgnoreCase("PROJECTILEPROTECTION")) {
                enchantName = "PROTECTION_PROJECTILE";
            } else if (enchantName.equalsIgnoreCase("FIREPROTECTION")) {
                enchantName = "PROTECTION_FIRE";
            } else if (enchantName.equalsIgnoreCase("BLASTPROTECTION")) {
                enchantName = "PROTECTION_EXPLOSIONS";
            } else if (enchantName.equalsIgnoreCase("PROTECTION")) {
                enchantName = "PROTECTION_ENVIRONMENTAL";
            } else if (enchantName.equalsIgnoreCase("FEATHERFALLING")) {
                enchantName = "PROTECTION_FALL";
            } else if (enchantName.equalsIgnoreCase("RESPIRATION")) {
                enchantName = "OXYGEN";
            } else if (enchantName.equalsIgnoreCase("AQUAAFFINITY")) {
                enchantName = "WATER_WORKER";
            } else if (enchantName.equalsIgnoreCase("UNBREAKING")) {
                enchantName = "DURABILITY";
            }
            Enchantment enchantment = Enchantment.getByName(enchantName);
            int level = -1;
            if (enchantment == null) {
                logger.log(Level.SEVERE, "Unrecognized enchantment: " + enchantName + " at config.yml path: " + path);
                hasErrors = true;
            }
            try {
                level = Integer.parseInt(parts[1]);
            } catch (NumberFormatException nfe) {
                logger.log(Level.SEVERE, "Unparseable integer: " + parts[1] + " at config.yml path: " + path);
                hasErrors = true;
            }
            if (level <= 0) {
                logger.log(Level.SEVERE, "Enchantment Level: " + parts[1] +
                        " at config.yml path: " + path + " needs to be greater than 0");
                hasErrors = true;
            }
            armorEnchants.put(enchantment, level);
        }
        return armorEnchants;
    }
    
    /**
     * Gets a map of enchants from Arraylist populated by loadEnchants(
     * @param index
     * @return Map of Enchants
     */
    public Map<Enchantment, Integer> getEnchantments(int index) {
        return this.armorEnchants.get(index);
    }

    /**
     * 
     * @param p Player object
     * @return true if OP or has permission "uniform.warden"
     */
    public static boolean isWarden(Player p) {
        if (p.isOp() || p.hasPermission("uniform.warden")) {
            return true;
        } else {
            return false;
        }
    }

}


me/menexia/guarduniform/GUMainListener.java
package me.menexia.guarduniform;

import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.entity.PlayerDeathEvent;
import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.event.player.PlayerDropItemEvent;
import org.bukkit.event.player.PlayerJoinEvent;
import org.bukkit.event.player.PlayerRespawnEvent;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.PlayerInventory;

public class GUMainListener implements Listener {
	private final GuardUniform plugin;
	public GUMainListener(GuardUniform plugin) {
		this.plugin = plugin;
		plugin.getServer().getPluginManager().registerEvents(this, plugin);
	}
	
	@EventHandler
	public void onGuardJoin(final PlayerJoinEvent event) {
		Player p = event.getPlayer();
		if (GuardUniform.isWarden(p)) return;
		GUArmorManager.checkAndSetUniform(plugin, p);
	}
	
	@EventHandler
	public void onGuardRespawn(final PlayerRespawnEvent event) {
		Player p = event.getPlayer();
		
		if (GuardUniform.isWarden(p)) {
			return;
		}
		
		PlayerInventory inv = p.getInventory();
		
		if (p.hasPermission("uniform.alphaguard")) {
			GUArmorManager.setAlphaGuardArmor(plugin, inv);
		} else if (p.hasPermission("uniform.guard")) {
			GUArmorManager.setGuardArmor(plugin, inv);
		} else if (p.hasPermission("uniform.trainee")) {
			GUArmorManager.setTraineeArmor(plugin, inv);
		}
		
	}
	
	@EventHandler
	public void onPlayerDeath(final PlayerDeathEvent event) {
		// Drop no chain armor and health potions if has permission
		Player p = event.getEntity();
		if (p.hasPermission("uniform.alphaguard")
				|| p.hasPermission("uniform.guard")
				|| p.hasPermission("uniform.trainee")) {
			GUArmorManager.noDropChainAndHealthPots(event);
			GUArmorManager.dropIronArmor(event);
		} else { // if not, just don't drop the chain.
			GUArmorManager.noDropChain(event);
		}
	}
	
	@SuppressWarnings("deprecation")
	@EventHandler
	public void onInventoryClick(final InventoryClickEvent event) {
		Player p = (Player) event.getWhoClicked();
		/*
		 * A Personal Anecdote:
		 * 
		 * Bukkit Bug:
		 * InventoryClickEvent is fired TWICE when opening the inventory
		 * in creative mode for the first time upon entering a server
		 * 
		 * Try un-commenting line 75. (v1.7)
		 */
//		p.sendMessage("ClickEvent fired!");
		
		
//		InventoryView view = event.getView();
//		Inventory ak = view.getTopInventory();
//		InventoryType type = view.getType();
//		p.sendMessage("" + event.getRawSlot() + " and " +  ak.getName() + "/" + ak.getTitle() + ": " + ak.getSize());
//		p.sendMessage(type.getDefaultTitle() + " " + type.name() + " " + type.toString());
		/*
		 * Result if Player Inventory:
		 * 18 and container.crafting/container.crafting: 5 (which is 4 crafting slots and 1 result slot, 0, 1, 2, 3, 4)
		 * Crafting CRAFTING CRAFTING
		 */
		
		
		if (GuardUniform.isWarden(p)) return;
		int slot = event.getSlot();
		ItemStack item = event.getCurrentItem();
		
		try {
			if (p.hasPermission("uniform.alphaguard")) {
				if (GUArmorManager.removeCheckAlphaGuard(plugin, p, slot, item)) { // is he trying to remove his armour?
					event.setCancelled(true);
				}
				else if (GUPotionManager.isRemovingPotionFromInventory(plugin, event)) { // is he not? then is he removing a potion from his inventory?
					event.setCancelled(true);
					p.updateInventory();
				}
			} else if (p.hasPermission("uniform.guard")) {
				 if (GUArmorManager.removeCheckGuard(plugin, p, slot, item)) {
					 event.setCancelled(true);
				 } else if (GUPotionManager.isRemovingPotionFromInventory(plugin, event)) {
					 event.setCancelled(true);
					 p.updateInventory();
				 }
			} else if (p.hasPermission("uniform.trainee")) {
				if (GUArmorManager.removeCheckTrainee(plugin, p, slot, item)) {
					event.setCancelled(true);
				} else if (GUPotionManager.isRemovingPotionFromInventory(plugin, event)) {
					event.setCancelled(true);
					p.updateInventory();
				}
			} else {
				// is he trying to add chain armour via shift click?
				if (GUArmorManager.addingArmourViaShiftClick(p, event.isShiftClick(), event.getCurrentItem())) {
					event.setCancelled(true);
					p.updateInventory();
				}
				// is he trying to add chain armour via normal click?
				else if (GUArmorManager.addingArmourViaNormalClick(p, slot, event.getCursor())) {
					event.setCancelled(true);
					p.updateInventory();
				}
			}
		} catch (NullPointerException nPE) {
			nPE.printStackTrace();
		}
	}
	
	@SuppressWarnings("deprecation")
	@EventHandler
	public void dropCheck(final PlayerDropItemEvent event) {
		Player p = event.getPlayer();
		if (GuardUniform.isWarden(p)) return;
		ItemStack itemStackDrop = event.getItemDrop().getItemStack();
		if (GUPotionManager.isDroppingHealthPotions(plugin, p, itemStackDrop)) {
			event.setCancelled(true);
			p.updateInventory();
		}
	}

}


me/menexia/guarduniform/GUPotionManager.java
package me.menexia.guarduniform;

import java.util.Arrays;

import org.bukkit.ChatColor;
import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.event.inventory.InventoryType;
import org.bukkit.event.inventory.InventoryType.SlotType;
import org.bukkit.inventory.ItemStack;

public class GUPotionManager {
	private final static InventoryType[] shiftClickInvents = {InventoryType.CHEST, InventoryType.ENDER_CHEST, InventoryType.DISPENSER, InventoryType.BREWING};
	
	public static boolean isRemovingPotionFromInventory(GuardUniform plugin, InventoryClickEvent event) {
		/*
		 * inventories that shift click potions into the upper inventory:
		 * chest, enderchest, dispenser, brewing stand
		 */
		Player p = (Player)event.getWhoClicked();
		ItemStack item;
		InventoryType inventoryType;
		// is it or is it not a shift click?
		if (event.isShiftClick()) {
			item = event.getCurrentItem();
			// is the item shift clicked a health potion?
			if (isHealthPotion(item)) {
				// is the inventoryType a chest, enderchest, dispenser or brewing stand?
				// (as they are possible inventories for potions to be shift clicked into)
				inventoryType = event.getView().getType();
				if (Arrays.asList(shiftClickInvents).contains(inventoryType)) {
					p.sendMessage(ChatColor.RED + "You cannot remove health potions from your inventory!");
					return true;
				}
			}
		} else {
			item = event.getCursor();
			// is the item on the cursor a health potion?
			if (isHealthPotion(item)) {
				// was the item not clicked within the quickbar?
				SlotType slotType = event.getSlotType();
				if (!slotType.equals(SlotType.QUICKBAR)) {
					// was the item clicked on the outside?
					inventoryType = event.getView().getType();
					if (slotType.equals(SlotType.OUTSIDE)
							|| ((!inventoryType.equals(InventoryType.PLAYER) || !inventoryType.equals(InventoryType.ANVIL)) && event.getRawSlot() < event.getView().getTopInventory().getSize())) {
						p.sendMessage(ChatColor.RED + "You cannot remove health potions from your inventory!");
						return true;
					}
				}
			}
		}
		return false;
	}
	
	private static boolean isHealthPotion(ItemStack item) {
		if (item != null && item.getType().equals(Material.POTION)) {
			if (Arrays.asList(GuardUniform.healPots).contains(item.getDurability())) {
				return true;
			}
		}
		return false;
	}
	
	public static boolean isDroppingHealthPotions(GuardUniform plugin, Player p, ItemStack itemStackDrop) {
		if (p.hasPermission("uniform.alphaguard") || p.hasPermission("uniform.guard") || p.hasPermission("uniform.trainee")) {
			if (isHealthPotion(itemStackDrop)) {
				p.sendMessage(ChatColor.RED + "You cannot remove health potions from your inventory!");
				return true;
			}
		}
		return false;
	}

}


me/menexia/guarduniform/GUDamageListener.java
package me.menexia.guarduniform;

import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.entity.EntityDamageEvent;
import org.bukkit.inventory.ItemStack;

public class GUDamageListener implements Listener {
	public GUDamageListener(GuardUniform plugin) {
        plugin.getServer().getPluginManager().registerEvents(this, plugin);
	}
	
	@EventHandler
	public void onDamageOnGuard(EntityDamageEvent event) {
		if (event.getEntity() instanceof Player) {
			Player p = (Player)event.getEntity();
			if (p.hasPermission("uniform.alphaguard") || p.hasPermission("uniform.guard") || p.hasPermission("uniform.trainee") || p.hasPermission("uniform.warden")) {
				ItemStack[] armor = p.getInventory().getArmorContents();	
				if (armor.length > 0) {
					for (int i = 0; i < armor.length; i++) {
						if (armor[i] != null) armor[i].setDurability((short)(-armor[i].getType().getMaxDurability()));
					}
				}
			}
		}
	}

}


me/menexia/guarduniform/GUArmorManager.java

package me.menexia.guarduniform;

import java.util.Arrays;
import java.util.Iterator;
import java.util.List;
import java.util.Random;

import org.bukkit.ChatColor;
import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.bukkit.event.entity.PlayerDeathEvent;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.PlayerInventory;

public class GUArmorManager {
	private static Random random = new Random(101);
	
	public static void checkAndSetUniform(GuardUniform plugin, Player p) {
		PlayerInventory inv = p.getInventory();
		// Perm checks...
		if (p.hasPermission("uniform.alphaguard")) {
			// Does he have the chain armour with the correct enchants?
			boolean hasCorrectArmor = GUArmorManager.hasAlphaGuardArmor(plugin, inv);
			if (!hasCorrectArmor) { // if he does not have the enchants
				GUArmorManager.moveNonChainArmorToInventory(p);
				GUArmorManager.setAlphaGuardArmor(plugin, inv);
			} else if (recallArmour(plugin)) { // Recalling the armour?
				GUArmorManager.clearEquippedChainArmor(p);
			}
		} else if (p.hasPermission("uniform.guard")) {
			boolean hasCorrectArmor = GUArmorManager.hasGuardArmor(plugin, inv);
			if (!hasCorrectArmor) {
				GUArmorManager.moveNonChainArmorToInventory(p);
				GUArmorManager.setGuardArmor(plugin, inv);
			} else if (recallArmour(plugin)) {
				GUArmorManager.clearEquippedChainArmor(p);
			}
		} else if (p.hasPermission("uniform.trainee")) {
			boolean hasCorrectArmor = GUArmorManager.hasTraineeArmor(plugin, inv);
			if (!hasCorrectArmor) {
				GUArmorManager.moveNonChainArmorToInventory(p);
				GUArmorManager.setTraineeArmor(plugin, inv);
			} else if (recallArmour(plugin)) {
				GUArmorManager.clearEquippedChainArmor(p);
			}
		} else { // not a guard and have chain?
			GUArmorManager.clearEquippedChainArmor(p);
		}
	}
	
	public static boolean hasAlphaGuardArmor(GuardUniform plugin, PlayerInventory inv) {
		return checkEnchArmor(plugin, inv, 0, 1, 2, 3);
	}
	
	public static boolean hasGuardArmor(GuardUniform plugin, PlayerInventory inv) {
		return checkEnchArmor(plugin, inv, 4, 5, 6, 7);
	}
	
	public static boolean hasTraineeArmor(GuardUniform plugin, PlayerInventory inv) {
		return checkEnchArmor(plugin, inv, 8, 9, 10, 11);
	}
	
	private static boolean checkEnchArmor(GuardUniform plugin, PlayerInventory inv, int helmet, int chestplate, int leggings, int boots) {
		if ((inv.getHelmet() != null && inv.getHelmet().getType().equals(Material.CHAINMAIL_HELMET) && inv.getHelmet().getEnchantments().equals(plugin.getEnchantments(helmet)))
				&& (inv.getChestplate() != null && inv.getChestplate().getType().equals(Material.CHAINMAIL_CHESTPLATE) && inv.getChestplate().getEnchantments().equals(plugin.getEnchantments(chestplate)))
				&& (inv.getLeggings() != null && inv.getLeggings().getType().equals(Material.CHAINMAIL_LEGGINGS) &&  inv.getLeggings().getEnchantments().equals(plugin.getEnchantments(leggings)))
				&& (inv.getBoots() != null && inv.getBoots().getType().equals(Material.CHAINMAIL_BOOTS) &&  inv.getBoots().getEnchantments().equals(plugin.getEnchantments(boots)))) {
			return true;
		} else {
			return false;
		}
	}
	
	private static void moveNonChainArmorToInventory(Player p) {
		PlayerInventory inv = p.getInventory();
		if (!GUArmorManager.hasEnoughEmptySlots(inv)) {
			GUMessages.sendPSA(p);
			return;
		}
		
		if (inv.getHelmet() != null && !inv.getHelmet().getType().equals(Material.CHAINMAIL_HELMET)) {
			inv.setItem(inv.firstEmpty(), inv.getHelmet());
		}
				
		if (inv.getChestplate() != null && inv.getChestplate().getType().equals(Material.CHAINMAIL_CHESTPLATE)) {
			inv.setItem(inv.firstEmpty(), inv.getChestplate());
		}
				
		if (inv.getLeggings() != null && !inv.getLeggings().getType().equals(Material.CHAINMAIL_LEGGINGS)) {
			inv.setItem(inv.firstEmpty(), inv.getLeggings());
		}

		if (inv.getBoots() != null && !inv.getBoots().getType().equals(Material.CHAINMAIL_BOOTS)) {
			inv.setItem(inv.firstEmpty(), inv.getBoots());
		}
		
	}
	
	private static boolean hasEnoughEmptySlots(PlayerInventory inv) {
		int count = 0;
		for (int i=0; i<36; i++) {
			if (inv.getItem(i) == null) count++;
		}
		if (count >= 4) return true; else return false;
	}
	
	public static void setAlphaGuardArmor(GuardUniform plugin, PlayerInventory inv) {
		setArmor(plugin, inv, 0, 1, 2, 3);
	}
	
	public static void setGuardArmor(GuardUniform plugin, PlayerInventory inv) {
		setArmor(plugin, inv, 4, 5, 6, 7);
	}
	
	public static void setTraineeArmor(GuardUniform plugin, PlayerInventory inv) {
		setArmor(plugin, inv, 8, 9, 10, 11);
	}
	
	private static void setArmor(GuardUniform plugin, PlayerInventory inv, int helmet, int chestplate, int leggings, int boots) {
		ItemStack helmeti = new ItemStack(Material.CHAINMAIL_HELMET, 1),
				chestplatei = new ItemStack(Material.CHAINMAIL_CHESTPLATE, 1),
				leggingsi = new ItemStack(Material.CHAINMAIL_LEGGINGS, 1),
				bootsi = new ItemStack(Material.CHAINMAIL_BOOTS, 1);
		helmeti.addUnsafeEnchantments(plugin.getEnchantments(helmet));
		chestplatei.addUnsafeEnchantments(plugin.getEnchantments(chestplate));
		leggingsi.addUnsafeEnchantments(plugin.getEnchantments(leggings));
		bootsi.addUnsafeEnchantments(plugin.getEnchantments(boots));
		inv.setHelmet(helmeti);
		inv.setChestplate(chestplatei);
		inv.setLeggings(leggingsi);
		inv.setBoots(bootsi);
	}
	
	private static boolean recallArmour(GuardUniform plugin) {
		return plugin.getConfig().getBoolean("recallArmor", false);
	}
	
	private static void clearEquippedChainArmor(Player p) {
		PlayerInventory inv = p.getInventory();
		ItemStack helm = inv.getHelmet();
		ItemStack chest = inv.getChestplate();
		ItemStack pants = inv.getLeggings();
		ItemStack boots = inv.getBoots();
		if (helm != null && helm.getType().equals(Material.CHAINMAIL_HELMET)) inv.setHelmet(null);
		if (chest != null && chest.getType().equals(Material.CHAINMAIL_CHESTPLATE)) inv.setChestplate(null);
		if (pants != null && pants.getType().equals(Material.CHAINMAIL_LEGGINGS)) inv.setLeggings(null);
		if (boots != null && boots.getType().equals(Material.CHAINMAIL_BOOTS)) inv.setBoots(null);
	}
	
	/*
	 * InventoryClickEvent ARMOUR calls
	 */
	
	public static boolean removeCheckAlphaGuard(GuardUniform plugin, Player p, int slot, ItemStack item) {
		return armorRemoveCheck(plugin, p, slot, item, 0, 1, 2, 3);
	}
	
	public static boolean removeCheckGuard(GuardUniform plugin, Player p, int slot, ItemStack item) {
		return armorRemoveCheck(plugin, p, slot, item, 4, 5, 6, 7);
	}

	public static boolean removeCheckTrainee(GuardUniform plugin, Player p, int slot, ItemStack item) {
		return armorRemoveCheck(plugin, p, slot, item, 8, 9, 10, 11);
	}
	
	private static boolean armorRemoveCheck(GuardUniform plugin, Player p, int slot, ItemStack item, int helmet, int chestplate, int leggings, int boots) {
		if ((slot==39 && item.getType().equals(Material.CHAINMAIL_HELMET) && item.getEnchantments().equals(plugin.getEnchantments(helmet)))
				|| (slot==38 && item.getType().equals(Material.CHAINMAIL_CHESTPLATE) && item.getEnchantments().equals(plugin.getEnchantments(chestplate)))
				|| (slot==37 && item.getType().equals(Material.CHAINMAIL_LEGGINGS) && item.getEnchantments().equals(plugin.getEnchantments(leggings)))
				|| (slot==36 && item.getType().equals(Material.CHAINMAIL_BOOTS) && item.getEnchantments().equals(plugin.getEnchantments(boots)))) {
			p.sendMessage(ChatColor.RED + "You cannot remove your uniform!");
			return true;
		}
		return false;
	}
	
	public static void noDropChainAndHealthPots(PlayerDeathEvent event) {
		List<ItemStack> items = event.getDrops();
		Iterator<ItemStack> it = items.iterator();
		while (it.hasNext()) {
			ItemStack item = it.next();
			if (!removeChain(it, item)) {
				removeInstaHealthPotion(it, item);
			}
		}
	}
	
	private static boolean removeChain(Iterator<ItemStack> it, ItemStack item) {
		if (item.getType().equals(Material.CHAINMAIL_HELMET)
				|| item.getType().equals(Material.CHAINMAIL_CHESTPLATE)
				|| item.getType().equals(Material.CHAINMAIL_LEGGINGS)
				|| item.getType().equals(Material.CHAINMAIL_BOOTS)) {
			it.remove();
			return true;
		}
		return false;
	}
	
	private static void removeInstaHealthPotion(Iterator<ItemStack> it, ItemStack item) {
		if (item.getType().equals(Material.POTION) && Arrays.asList(GuardUniform.healPots).contains(item.getDurability())) {
			it.remove();
		}
	}
	
	public static void dropIronArmor(PlayerDeathEvent event) {
		List<ItemStack> items = event.getDrops();
		ItemStack helmet = new ItemStack(Material.IRON_HELMET),
				chestplate = new ItemStack(Material.IRON_CHESTPLATE),
				leggings = new ItemStack(Material.IRON_LEGGINGS),
				boots = new ItemStack(Material.IRON_BOOTS);
		if (random.nextBoolean()) helmet.setDurability((short)random.nextInt(83));
		else helmet.setDurability((short)(random.nextInt(83)+83));
		if (random.nextBoolean()) chestplate.setDurability((short)random.nextInt(120));
		else chestplate.setDurability((short)(random.nextInt(120)+120));
		if (random.nextBoolean()) leggings.setDurability((short)random.nextInt(113));
		else leggings.setDurability((short)(random.nextInt(113)+113));
		if (random.nextBoolean()) boots.setDurability((short)random.nextInt(98));
		else boots.setDurability((short)(random.nextInt(98)+98));
		
		items.add(helmet);
		items.add(chestplate);
		items.add(leggings);
		items.add(boots);
	}
	
	public static void noDropChain(PlayerDeathEvent event) {
		List<ItemStack> items = event.getDrops();
		Iterator<ItemStack> it = items.iterator();
		while (it.hasNext()) {
			ItemStack item = it.next();
			removeChain(it, item);
		}
	}
	
	public static boolean addingArmourViaShiftClick(Player p, boolean isShiftClick, ItemStack armor) {
		if (armor != null) {
			if (isShiftClick) {
				Material armorMat = armor.getType();
				if (armorMat.equals(Material.CHAINMAIL_HELMET) || armorMat.equals(Material.CHAINMAIL_CHESTPLATE) || armorMat.equals(Material.CHAINMAIL_LEGGINGS) || armorMat.equals(Material.CHAINMAIL_BOOTS)) {
					p.sendMessage(ChatColor.RED + "You cannot equip chain armor!");
					return true;
				}
			}
		}
		return false;
	}

	public static boolean addingArmourViaNormalClick(Player p, int slot, ItemStack cursor) {
		if ((slot==39 && cursor.getType().equals(Material.CHAINMAIL_HELMET))
				|| (slot==38 && cursor.getType().equals(Material.CHAINMAIL_CHESTPLATE))
				|| (slot==37 && cursor.getType().equals(Material.CHAINMAIL_LEGGINGS))
				|| (slot==36 && cursor.getType().equals(Material.CHAINMAIL_BOOTS))) {
			p.sendMessage(ChatColor.RED + "You cannot equip chain armor!");
			return true;
		}
		return false;
	}
	
}
